### PR TITLE
Fix flexattention not realizing inputs before lowering (also refactored runtime estimation)

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126641
* #126446
* __->__ #126615

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/sizevars.py [hunk 1]
```diff
@@ -449,6 +449,7 @@ def size_hint(self, expr: Expr, *, fallback: Optional[int] = None) -> int:
         try:
             return int(out)
         except Exception:
+            breakpoint()
```
Comment (Skylion007) [inline torch/_inductor/sizevars.py:452]: I don't think you mean to leave this in.

File: torch/_inductor/sizevars.py [hunk 2]
```diff
@@ -449,6 +449,7 @@ def size_hint(self, expr: Expr, *, fallback: Optional[int] = None) -> int:
...
             log.debug("failed on: %s", out)
```
Comment (Skylion007) [inline torch/_inductor/sizevars.py:453]:
```suggestion
            log.debug("failed on: %s", out, exc_info=True)
```

**Review by yanboliang — APPROVED** [review]

File: test/inductor/test_control_flow.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 import itertools
 
 import torch
+import torch._dynamo.testing
```
Comment (drisspg) [inline test/inductor/test_control_flow.py:5]: ?

File: test/inductor/test_flex_attention.py [hunk 4]
```diff
@@ -645,6 +646,50 @@ def f(q, k1, k2, k3, v1, v2, v3):
...
+        q, k, v = (torch.randn(1, 8, 1024, 64, device="cuda") for _ in range(3))
+        metrics.reset()
+        f(q, k, v)
+        accessed_bytes = 1 * 8 * 1024 * 64 * torch.float32.itemsize
+        num_accesses = 4  # q, k, v reads, one output.
+        # TODO: Get rid of this fudge factor
+        # We need this fudge factor for now, since
+        # 1. For some reason we materialize the output of the attention unnecessarily (it's related to the mutation somehow)
+        # 2. We also write the extraneous logsumexp
+        num_accesses += 2
```
Comment (drisspg) [inline test/inductor/test_flex_attention.py:690]: the logsum exp is  1 * 8 * 1024 though right, should we make this more restrictive?

**Review by drisspg — APPROVED** [review]
The flex changes look good, the other looks like a lot of code motion

**Review by xmfan — COMMENTED** [review]
~~could we cleanup the now unused globals from torch/_inductor/metrics.py and in reset()~~

**Review by xmfan — APPROVED** [review]

File: test/inductor/test_control_flow.py (hunk 3 above)
Comment (Chillee) [reply test/inductor/test_control_flow.py:5]: This test file requires `torch._dynamo.testing` to be imported.

Somehow one of my import removals prevented the existing path, so i needed to readd it here.

https://github.com/pytorch/pytorch/blob/d133d5403bca0856429f12ab0e49c4b9c89d1bec/test/inductor/test_control_flow.py#L182

File: test/inductor/test_flex_attention.py (hunk 4 above)
Comment (Chillee) [reply test/inductor/test_flex_attention.py:690]: yeah in principle, but I would hope that we fix both of these issues soonish :)

**Comment by DanilBaibak** [comment]
Hi @Chillee! Sorry, I need to revert your PR because it breaks the internal a lot of inductor tests (triton_kernels and pattern_matcher).  Here you can find more information about the issues - [D57619189](https://www.internalfb.com/diff/D57619189).

**Pull request closed without merging**